src/compartment.c

```diff
@@ -0,0 +1,526 @@
+                        }
+                        default:
+                            assert(false);
+                    }
+                    found += 1;
+                    printf("Found `main` func at %p.\n", (void*) new_comp->entry_point);
+                }
+                else
+                {
+                    for (size_t i = 0; i < MAX_INTERCEPT_COUNT; ++i)
+                    {
+                        if (!strcmp(comp_intercept_funcs[i].func_name, &comp_strtb[curr_sym.st_name]))
+                        {
+                            comp_add_intercept(new_comp, curr_sym.st_value, comp_intercept_funcs[i]);
+                            found += 1;
+                            break;
+                        }
+                    }
+                }
+                if (found == to_find)
+                {
+                    break;
+                }
+            }
+            free(comp_symtb);
+            free(comp_strtb);
+        }
+        // TODO still need relas check or consider only static executables?
+        else if (comp_symtb_hdr.sh_type == SHT_RELA) // TODO change name && consider SH_REL
+        {
+            if (comp_symtb_hdr.sh_info == 0) // TODO better identify the plt relocation section
+            {
+                continue;
+            }
+
+            new_comp->relas_cnt = comp_symtb_hdr.sh_size / sizeof(Elf64_Rela);
+            new_comp->relas = (uintptr_t*) malloc(new_comp->relas_cnt * sizeof(uintptr_t));
+            Elf64_Rela* comp_relas = (Elf64_Rela*) malloc(comp_symtb_hdr.sh_size);
+            pread_res = pread((int) new_comp->fd, comp_relas, comp_symtb_hdr.sh_size, comp_symtb_hdr.sh_offset);
+            assert(pread_res != -1);
+
+            for (size_t j = 0; j < new_comp->relas_cnt; ++j)
+            {
+                new_comp->relas[j] = comp_relas[j].r_offset;
+                /*comp_relas[j].r_offset += new_comp->base;*/
+                /*uintptr_t old_plt_val = (uintptr_t) *((void**) comp_relas[j].r_offset);*/
+                /*old_plt_val += new_comp->base;*/
+            }
+
+            free(comp_relas);
+        }
+        else
+        {
+            continue;
+        }
+    }
+    /*assert(found == to_find);*/
+
+    comp_register_ddc(new_comp);
+    comp_print(new_comp);
+    return new_comp;
+}
+
+void
+comp_register_ddc(struct Compartment* new_comp)
+{
+    void* __capability new_ddc = cheri_address_set(manager_ddc, new_comp->base);
+    new_ddc = cheri_bounds_set(new_ddc, new_comp->size + new_comp->scratch_mem_size + new_comp->scratch_mem_stack_size);
+    // TODO bounds double-check
+    new_comp->ddc = new_ddc;
+}
+
+void
+comp_add_intercept(struct Compartment* new_comp, uintptr_t intercept_target, struct func_intercept intercept_data)
+{
+    printf("Found `%s` func at %p.\n", intercept_data.func_name, (void*) intercept_target);
+    int offset = (intercept_data.redirect_func - intercept_target) / 4;
+    assert(offset < (1 << 27));
```

Comment:
Why this assert?

Reply:
This is something more for me, since the specific instruction I'm building has 6 bits to determine the instruction type, then 26 bits for the parameter. Just making sure I do correct things, and check myself. I'll be replacing this function shortly anyway, as I need a more generic way of creating instructions.

Reply:
Ah, I'm with you. I'd perhaps suggest that `assert((offset & ~0b1111111) == 0)` might be clearer, but it's not a big deal.

Reply:
As I said, this is getting changed very soon, so I'll keep your suggestion in mind. Unsure, at first glance, if what you're proposing is equivalent, but I'll try to make it less obscure as just a constant (and add a comment as well).